fs2/src/main/scala/laserdisc/fs2/PromiseMapper.scala

```diff
@@ -1,28 +1,34 @@
 package laserdisc
 package fs2
 
-import _root_.fs2.{Scheduler, async}
-import cats.effect.Effect
+import cats.effect.concurrent.Deferred
+import cats.effect.{Concurrent, Timer}
+import cats.effect.syntax.concurrent._
+import cats.syntax.flatMap._
+import cats.syntax.monadError._
 import shapeless.Poly1
 
-import scala.concurrent.ExecutionContext
+import scala.concurrent.TimeoutException
 import scala.concurrent.duration.FiniteDuration
 
 object PromiseMapper extends Poly1 {
-  implicit def mkOne[F[_], A](implicit ec: ExecutionContext, s: Scheduler) = at[(Protocol.Aux[A], Effect[F])] {
-    case (protocol, effect) =>
-      (queueAndFiniteDuration: (Queue[F], FiniteDuration)) =>
-        queueAndFiniteDuration match {
-          case (queue: Queue[F], timeout: FiniteDuration) =>
-            effect.flatMap(async.promise[F, Maybe[Maybe[A]]](effect, ec)) { promise =>
-              effect.flatMap(queue.enqueue1(Request(protocol, promise.complete))) { _ =>
-                effect.flatMap(promise.timedGet(timeout, s)) {
-                  case None                  => effect.raiseError[Maybe[A]](RequestTimedOut(protocol))
-                  case Some(Left(throwable)) => effect.raiseError[Maybe[A]](throwable)
-                  case Some(Right(response)) => effect.pure(response)
+  implicit def mkOne[F[_]: Concurrent: Timer, A] = at[Protocol.Aux[A]] {
+    protocol => (queueAndDuration: (Queue[F], FiniteDuration)) =>
+      queueAndDuration match {
+        case (queue: Queue[F], duration: FiniteDuration) =>
+          Deferred[F, Maybe[Maybe[A]]].flatMap { promise =>
```

Comment:
What do you think about ?
```scala
Deferred[F, Throwable | (Throwable | A)].flatMap { promise =>
```

Reply:
so `type Maybe[+A] = Either[Throwable, A]` while IIRC you have defined elsewhere `type |[+A, +B] = Either[A, B]`. To me too `A | (A | B)` reads better, but we're now repeating the fixed `A` over and over, with `Maybe`, although more verbose than `|` we don't. I wish [SIP-36](https://docs.scala-lang.org/sips/adding-prefix-types.html) could fix this, but it's concerned only on lifting to the type level the same restrictions on `unary_` that we have at the value level (so `+`, `-`, `~`, `!`)

Reply:
Actually, the whole dance of Either[Either[A]] can probably be removed, esp since I was raising the error in case of a failure